File: platform/Emscripten/src/benchmark_report_formatter.js
const { formatSize } = require("./benchmark_case");

function padLeft(value, width) {
  return String(value).padStart(width, " ");
}

function padRight(value, width) {
  return String(value).padEnd(width, " ");
}

function fixed(value) {
  return value.toFixed(3);
}

function formatRows(rows, title) {
  const lines = [];
  lines.push("");
  lines.push(title);
  lines.push(
    `${padRight("Binding", 16)}${padRight("Operation", 32)}${padRight("Case", 12)}` +
      `${padLeft("Iterations", 12)}${padLeft("Elements", 12)}${padLeft("Avg (ms)", 14)}` +
      `${padLeft("Min (ms)", 14)}${padLeft("Max (ms)", 14)}${padLeft("Bytes", 14)}`,
  );

  for (const row of rows) {
    lines.push(
      `${padRight(row.binding, 16)}${padRight(row.operation, 32)}${padRight(row.case.name, 12)}` +
        `${padLeft(row.case.iterations, 12)}${padLeft(row.elementCount, 12)}` +
        `${padLeft(fixed(row.measurement.averageMs), 14)}${padLeft(fixed(row.measurement.minMs), 14)}` +
        `${padLeft(fixed(row.measurement.maxMs), 14)}${padLeft(row.case.targetBytes, 14)}`,
    );
  }

  return lines.join("\n");
}

function formatReport({ modulePath, cases, modelCounts, u32ElementCounts, rows, sink }) {
  const lines = [];
  lines.push("ffi-binary-bench Emscripten Node.js benchmark");
  lines.push("Payload families: BinaryModel, u32 baseline");
  lines.push("Wire formats: little-endian BinaryModel records, raw contiguous u32 bytes");
  lines.push(`Module: ${modulePath}`);
  lines.push(`Node.js: ${process.version}`);
  lines.push("Operations:");
  lines.push("  BinaryModel: JS objects -> compact bytes -> wasm C API decode");
  lines.push("  BinaryModel: wasm C API encode -> compact bytes -> JS objects");
  lines.push("  u32: Uint32Array -> compact bytes -> wasm C API decode");
  lines.push("  u32: wasm C API encode -> compact bytes -> Uint32Array/view");
  lines.push("");
  lines.push("Datasets");

  for (let index = 0; index < cases.length; index += 1) {
    const benchmarkCase = cases[index];
    lines.push(
      `  ${padRight(benchmarkCase.name, 8)} size=${padRight(formatSize(benchmarkCase.targetBytes), 10)}` +
        ` iterations=${benchmarkCase.iterations}` +
        ` binary_models=${modelCounts[index]}` +
        ` u32_elements=${u32ElementCounts[index]}`,
    );
  }

  lines.push(formatRows(rows.filter((row) => row.payloadFamily === "BinaryModel"), "BinaryModel"));
  lines.push(formatRows(rows.filter((row) => row.payloadFamily === "u32"), "u32 Baseline"));
  lines.push("");
  lines.push(`Sink: ${sink}`);
  return `${lines.join("\n")}\n`;
}

function jsonRows(rows) {
  return {
    rows: rows.map((row) => ({
      payload_family: row.payloadFamily,
      binding: row.binding,
      operation: row.operation,
      case_name: row.case.name,
      payload_bytes: row.case.targetBytes,
      element_count: row.elementCount,
      iterations: row.case.iterations,
      average_ms: row.measurement.averageMs,
      min_ms: row.measurement.minMs,
      max_ms: row.measurement.maxMs,
    })),
  };
}

module.exports = {
  formatReport,
  jsonRows,
};

File: platform/Android/app/src/main/java/com/ffibb/android/BenchmarkReportFormatter.java
package com.ffibb.android;

import java.util.List;
import java.util.Locale;

final class BenchmarkReportFormatter {
  private BenchmarkReportFormatter() {
  }

  static String format(
      List<BenchmarkCase> cases,
      List<Integer> modelCounts,
      List<Integer> u32ElementCounts,
      List<BenchmarkRow> rows) {
    StringBuilder report = new StringBuilder();

    report.append("ffi-binary-bench Android benchmark\n");
    report.append("Payload families: BinaryModel, u32 baseline\n");
    report.append("Wire formats: little-endian BinaryModel records, raw contiguous u32 bytes\n");
    report.append("Operations:\n");
    report.append("  BinaryModel: List<BinaryModel> -> compact bytes -> JNI -> C API decode\n");
    report.append("  BinaryModel: C API encode -> JNI -> compact bytes -> List<BinaryModel>\n");
    report.append("  u32: int[] -> compact bytes -> JNI -> C API decode\n");
    report.append("  u32: C API encode -> JNI -> compact bytes -> int[]\n");
    report.append("\nDatasets\n");

    for (int index = 0; index < cases.size(); index++) {
      BenchmarkCase benchmarkCase = cases.get(index);
      report.append(String.format(
          Locale.US,
          "  %-8s size=%-10s iterations=%d binary_models=%d u32_elements=%d\n",
          benchmarkCase.name,
          formatSize(benchmarkCase.targetBytes),
          benchmarkCase.iterations,
          modelCounts.get(index),
          u32ElementCounts.get(index)));
    }

    appendRowsForPayloadFamily(report, rows, "BinaryModel", "BinaryModel");
    appendRowsForPayloadFamily(report, rows, "u32", "u32 Baseline");

    report.append("\nSink: ").append(BenchmarkSink.value()).append('\n');
    return report.toString();
  }

  private static void appendRowsForPayloadFamily(
      StringBuilder report,
      List<BenchmarkRow> rows,
      String payloadFamily,
      String title) {
    report.append("\n").append(title).append("\n");
    report.append(String.format(
        Locale.US,
      "%-31s %-35s %-12s %12s %12s %14s %14s %14s %14s\n",
        "Binding",
        "Operation",
        "Case",
        "Iterations",
        "Elements",
        "Avg (ms)",
        "Min (ms)",
        "Max (ms)",
        "Bytes"));

    for (BenchmarkRow row : rows) {
      if (!row.payloadFamily.equals(payloadFamily)) {
        continue;
      }

      report.append(String.format(
          Locale.US,
          "%-31s %-35s %-12s %12d %12d %14.3f %14.3f %14.3f %14d\n",
          row.binding,
          row.operation,
          row.caseName,
          row.iterations,
          row.elementCount,
          row.measurement.averageMs,
          row.measurement.minMs,
          row.measurement.maxMs,
          row.targetBytes));
    }
  }

  private static String formatSize(int bytes) {
    double kib = 1024.0;
    double mib = 1024.0 * 1024.0;
    if (bytes >= mib) {
      return String.format(Locale.US, "%.2f MiB", bytes / mib);
    }
    if (bytes >= kib) {
      return String.format(Locale.US, "%.2f KiB", bytes / kib);
    }
    return bytes + " B";
  }
}
